diff --git a/blackwell/5_tcgen05_mxfp8_2cta_256n_warp_specialized/tcgen05_mxfp8_2cta_256n_warp_specialized.cu b/blackwell/5_tcgen05_mxfp8_2cta_256n_warp_specialized/tcgen05_mxfp8_2cta_256n_warp_specialized.cu
index 3c8a4e1..18e623a 100644
--- a/blackwell/5_tcgen05_mxfp8_2cta_256n_warp_specialized/tcgen05_mxfp8_2cta_256n_warp_specialized.cu
+++ b/blackwell/5_tcgen05_mxfp8_2cta_256n_warp_specialized/tcgen05_mxfp8_2cta_256n_warp_specialized.cu
@@ -562,8 +562,7 @@ void producer_warp(
     constexpr int SMEM_SFB_SIZE = BN * SF_BK;
 
     int tma_smem_buf = 0;
-    int smem_empty_parity[QUEUE_SIZE];
-    for (int i = 0; i < QUEUE_SIZE; i++) smem_empty_parity[i] = 0;
+    int smem_empty_mbar_parity = 0; // flips every QUEUE_SIZE iterations when we wraparound and re-use smem buffers
 
     for (int group_id = start_group_id; group_id < total_groups; group_id += num_groups) {
         const int bid = group_id * CTA_GROUP_SIZE + (start_bid % CTA_GROUP_SIZE);
@@ -577,8 +576,7 @@ void producer_warp(
         for (int block_k_idx = 0; block_k_idx < num_blocks_k; block_k_idx++) {
             if (block_k_idx >= QUEUE_SIZE || group_id > start_group_id)
             {
-                mbarrier_wait_parity(smem_empty_mbar_addr + tma_smem_buf * sizeof(uint64_t), smem_empty_parity[tma_smem_buf]);
-                smem_empty_parity[tma_smem_buf] ^= 1;
+                mbarrier_wait_parity(smem_empty_mbar_addr + tma_smem_buf * sizeof(uint64_t), smem_empty_mbar_parity);
             }
 
             // smem offsets for next A/B tiles - layout: [A, B, SFA, SFB]
@@ -640,6 +638,8 @@ void producer_warp(
                 );
             }
             tma_smem_buf = (tma_smem_buf + 1) % QUEUE_SIZE;
+            if (tma_smem_buf == 0)
+                smem_empty_mbar_parity ^= 1;
         }
     }
 }
